File: src/main/java/org/infinitystudio/singularity/api/resource/ResourcePacket.scala
package org.infinitystudio.singularity.api.resource

import scala.collection.mutable

/**
 * @author Blealtan
 */
class ResourcePacket(resources: Array[Resource]) {
  private val packet : mutable.Map[Resource.SourceType, Resource] = mutable.Map()
  if (resources != null) resources.foreach(r => packet getOrElseUpdate (r.getType, r))

  def +=(resourcePacket: ResourcePacket) : ResourcePacket = {
    resourcePacket.packet.foreach(r => {
      val tmp = packet(r._1)
      tmp.setQuantity(packet(r._1).getQuantity + r._2.getQuantity)
      packet update (r._1, tmp)
    })
    this
  }

  def +(resourcePacket: ResourcePacket) : ResourcePacket = {
    val ret = new ResourcePacket(packet.values.toArray)
    ret += resourcePacket
  }

  def -=(resourcePacket: ResourcePacket) : ResourcePacket = {
    resourcePacket.packet.foreach(r => {
      val tmp = packet(r._1)
      tmp.setQuantity(packet(r._1).getQuantity - r._2.getQuantity)
      packet update (r._1, tmp)
    })
    this
  }

  def -(resourcePacket: ResourcePacket) : ResourcePacket = {
    val ret = new ResourcePacket(packet.values.toArray)
    ret -= resourcePacket
  }

  def get(r: Resource.SourceType) = packet get r

  def set(r: Resource.SourceType, res: Resource) = packet update(r, res)
}

File: src/main/java/org/infinitystudio/singularity/api/resource/ResourcePacket.java
package org.infinitystudio.singularity.api.resource;

import java.util.EnumMap;

/**
 * @author Blealtan
 */
public class ResourcePacket {
	private EnumMap<Resource.ResourceType, Resource> resources;

	/**
	 * Initialise the resource packet.
	 *
	 * @param resources The argument list of resources to be moved into the packet.
	 */
	public ResourcePacket(Resource... resources) {
		this.resources = new EnumMap<Resource.ResourceType, Resource>(Resource.ResourceType.class);
		for (Resource r : resources)
			if (!this.resources.containsKey(r.getType()))
				this.resources.put(r.getType(), r);
	}

	/**
	 * Set the resources in the packet if one type is specified in argument list.
	 * Not mentioned resource type will not be influenced.
	 *
	 * @param resources The argument list of resources to be moved into the packet.
	 */
	public ResourcePacket setResource(Resource... resources) {
		for (Resource r : resources)
			if (!this.resources.containsKey(r.getType()))
				this.resources.put(r.getType(), r);
		return this;
	}

	/**
	 * Get resource information for specified source type.
	 *
	 * @param resourceType The specified type.
	 * @return The wanted resource.
	 */
	public Resource getResource(Resource.ResourceType resourceType) {
		return resources.get(resourceType);
	}

	/**
	 * Append a resource packet to this one.
	 *
	 * @param resourcePacket The packet to be appended here.
	 * @return this
	 */
	public ResourcePacket appendResource(ResourcePacket resourcePacket) {
		for (Resource r : resourcePacket.resources.values()) {
			if (!this.resources.containsKey(r.getType()))
				this.resources.put(r.getType(), r);
			else
				this.resources.put(r.getType(), r.setQuantity(this.resources.get(r.getType()).getQuantity() + r.getQuantity()));
		}
		return this;
	}

	/**
	 * Remove a resource packet from this one.
	 *
	 * @param resourcePacket The packet to be removed from here.
	 * @return At most `this`. If not enough, null.
	 */
	public ResourcePacket removeResource(ResourcePacket resourcePacket) {
		ResourcePacket res = this;
		for (Resource r : resourcePacket.resources.values()) {
			if (!this.resources.containsKey(r.getType()))
				this.resources.put(r.getType(), r);
			else {
				int toset = this.resources.get(r.getType()).getQuantity() - r.getQuantity();
				if (toset < 0) {
					res = null;
					toset = 0;
				}
				this.resources.put(r.getType(), r.setQuantity(toset));
			}
		}
		return res;
	}
}
